File: ios/Runner/Tajweed/TajweedLiveCaptureDump.swift
import Foundation

/// Debug dump of Tajweed captures for cross-platform comparison.
/// Always overwrites `last.wav` + `last_stages.json` (compatibility), and also
/// archives a uniquely named `{surah}_{ayah}_attempt_NNNN.{wav,json}` so history
/// is never overwritten.
enum TajweedLiveCaptureDump {
  static var enabled: Bool = true

  private static let lock = NSLock()

  static var directory: URL {
    let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
    return docs.appendingPathComponent("TajweedLiveCompare", isDirectory: true)
  }

  static func dump(pcm: [Float], stages: [String: Any]) {
    guard enabled else { return }
    lock.lock()
    defer { lock.unlock() }

    let fm = FileManager.default
    let dir = directory
    do {
      try fm.createDirectory(at: dir, withIntermediateDirectories: true)

      // Compatibility path (existing tooling / pull scripts).
      let lastWav = dir.appendingPathComponent("last.wav")
      let lastJson = dir.appendingPathComponent("last_stages.json")
      try writeWavMono16k(pcm: pcm, to: lastWav)

      let (surah, ayah) = parseRef(stages)
      let attempt = nextAttemptIndex(in: dir, surah: surah, ayah: ayah)
      let basename = String(format: "%d_%d_attempt_%04d", surah, ayah, attempt)
      let archived = enrichStages(
        stages,
        surah: surah,
        ayah: ayah,
        attemptIndex: attempt,
        basename: basename
      )
      let jsonData = try JSONSerialization.data(
        withJSONObject: archived,
        options: [.prettyPrinted, .sortedKeys]
      )
      try jsonData.write(to: lastJson, options: .atomic)

      let archiveWav = dir.appendingPathComponent("\(basename).wav")
      let archiveJson = dir.appendingPathComponent("\(basename).json")
      try writeWavMono16k(pcm: pcm, to: archiveWav)
      try jsonData.write(to: archiveJson, options: .atomic)

      NSLog(
        "[TajweedLiveCompare] wrote last.wav + %@ (%d samples) attempt=%d",
        basename,
        pcm.count,
        attempt
      )
    } catch {
      NSLog("[TajweedLiveCompare] dump failed: \(error.localizedDescription)")
    }
  }

  private static func parseRef(_ stages: [String: Any]) -> (Int, Int) {
    if let s = stages["surah"] as? Int, let a = stages["ayah"] as? Int {
      return (s, a)
    }
    if let s = stages["surah"] as? NSNumber, let a = stages["ayah"] as? NSNumber {
      return (s.intValue, a.intValue)
    }
    let ref = (stages["ref"] as? String) ?? ""
    let parts = ref.split(separator: ":")
    let surah = parts.count >= 1 ? Int(parts[0]) ?? 0 : 0
    let ayah = parts.count >= 2 ? Int(parts[1]) ?? 0 : 0
    return (surah, ayah)
  }

  /// Next 1-based attempt index for this surah:ayah (never reuses a basename).
  private static func nextAttemptIndex(in dir: URL, surah: Int, ayah: Int) -> Int {
    let prefix = String(format: "%d_%d_attempt_", surah, ayah)
    let fm = FileManager.default
    guard let names = try? fm.contentsOfDirectory(atPath: dir.path) else { return 1 }
    var maxN = 0
    for name in names {
      guard name.hasPrefix(prefix), name.hasSuffix(".wav") else { continue }
      let mid = name.dropFirst(prefix.count).dropLast(4) // strip .wav
      if let n = Int(mid) {
        maxN = max(maxN, n)
      }
    }
    return maxN + 1
  }

  private static func enrichStages(
    _ stages: [String: Any],
    surah: Int,
    ayah: Int,
    attemptIndex: Int,
    basename: String
  ) -> [String: Any] {
    var out = stages
    out["surah"] = surah
    out["ayah"] = ayah
    out["attemptIndex"] = attemptIndex
    out["attemptBasename"] = basename
    out["expectedRaw"] = stages["expected"] ?? stages["originalExpectedAyah"] ?? ""
    out["hypRaw"] = stages["hypothesis"] ?? stages["originalAsrHypothesis"] ?? ""
    out["accuracy"] = stages["wordAccuracy"] ?? 0.0
    out["ops"] = stages["lexicalOps"] ?? []
    let modelInfo = stages["modelInfo"] as? [String: Any]
    out["bucket"] = modelInfo?["bucketOrFixedT"] ?? stages["bucketT"] ?? 0
    out["coldLoad"] = modelInfo?["modelWasColdLoad"] ?? false
    out["timestamp"] = ISO8601DateFormatter().string(from: Date())
    return out
  }

  /// PCM float [-1,1] → 16-bit mono WAV @ 16 kHz.
  static func writeWavMono16k(pcm: [Float], to url: URL) throws {
    let sampleRate: Int = 16_000
    var int16 = [Int16](repeating: 0, count: pcm.count)
    for i in 0..<pcm.count {
      let clamped = max(-1.0, min(1.0, Double(pcm[i])))
      int16[i] = Int16((clamped * Double(Int16.max)).rounded())
    }
    let dataSize = UInt32(int16.count * 2)
    var header = Data()
    func append(_ s: String) { header.append(contentsOf: s.utf8) }
    func appendU32(_ v: UInt32) {
      var le = v.littleEndian
      withUnsafeBytes(of: &le) { header.append(contentsOf: $0) }
    }
    func appendU16(_ v: UInt16) {
      var le = v.littleEndian
      withUnsafeBytes(of: &le) { header.append(contentsOf: $0) }
    }
    append("RIFF")
    appendU32(36 + dataSize)
    append("WAVE")
    append("fmt ")
    appendU32(16)
    appendU16(1) // PCM
    appendU16(1) // mono
    appendU32(UInt32(sampleRate))
    appendU32(UInt32(sampleRate * 2))
    appendU16(2) // block align
    appendU16(16) // bits
    append("data")
    appendU32(dataSize)
    var body = Data(count: int16.count * 2)
    body.withUnsafeMutableBytes { raw in
      let dest = raw.bindMemory(to: Int16.self)
      for i in 0..<int16.count {
        dest[i] = int16[i].littleEndian
      }
    }
    try (header + body).write(to: url, options: .atomic)
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/TajweedLiveCaptureDump.kt
package com.rnr.deenfocus.tajweed

import android.content.Context
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import org.json.JSONArray
import org.json.JSONObject

/**
 * Debug dump of Tajweed captures for cross-platform comparison.
 * Always overwrites `last.wav` + `last_stages.json` (compatibility), and also
 * archives a uniquely named `{surah}_{ayah}_attempt_NNNN.{wav,json}` so history
 * is never overwritten.
 */
object TajweedLiveCaptureDump {
    private const val TAG = "TajweedLiveCompare"

    @JvmField
    var enabled: Boolean = true

    private val lock = ReentrantLock()

    fun directory(context: Context): File =
        File(context.getExternalFilesDir(null), "TajweedLiveCompare").also { it.mkdirs() }

    fun dump(context: Context, pcm: FloatArray, stages: Map<String, Any?>) {
        if (!enabled) return
        lock.withLock {
            val dir = directory(context)
            try {
                val lastWav = File(dir, "last.wav")
                val lastJson = File(dir, "last_stages.json")
                writeWavMono16k(pcm, lastWav)

                val (surah, ayah) = parseRef(stages)
                val attempt = nextAttemptIndex(dir, surah, ayah)
                val basename = String.format(Locale.US, "%d_%d_attempt_%04d", surah, ayah, attempt)
                val archived = enrichStages(stages, surah, ayah, attempt, basename)
                val jsonText = (mapToJson(archived) as JSONObject).toString(2)
                lastJson.writeText(jsonText)

                val archiveWav = File(dir, "$basename.wav")
                val archiveJson = File(dir, "$basename.json")
                writeWavMono16k(pcm, archiveWav)
                archiveJson.writeText(jsonText)

                Log.i(
                    TAG,
                    "wrote last.wav + $basename (${pcm.size} samples) attempt=$attempt",
                )
            } catch (e: Exception) {
                Log.e(TAG, "dump failed: ${e.message}", e)
            }
        }
    }

    private fun parseRef(stages: Map<String, Any?>): Pair<Int, Int> {
        val surahDirect = (stages["surah"] as? Number)?.toInt()
        val ayahDirect = (stages["ayah"] as? Number)?.toInt()
        if (surahDirect != null && ayahDirect != null) return surahDirect to ayahDirect
        val ref = stages["ref"]?.toString().orEmpty()
        val parts = ref.split(":")
        val surah = parts.getOrNull(0)?.toIntOrNull() ?: 0
        val ayah = parts.getOrNull(1)?.toIntOrNull() ?: 0
        return surah to ayah
    }

    private fun nextAttemptIndex(dir: File, surah: Int, ayah: Int): Int {
        val prefix = String.format(Locale.US, "%d_%d_attempt_", surah, ayah)
        var maxN = 0
        val names = dir.list() ?: return 1
        for (name in names) {
            if (!name.startsWith(prefix) || !name.endsWith(".wav")) continue
            val mid = name.removePrefix(prefix).removeSuffix(".wav")
            val n = mid.toIntOrNull() ?: continue
            if (n > maxN) maxN = n
        }
        return maxN + 1
    }

    private fun enrichStages(
        stages: Map<String, Any?>,
        surah: Int,
        ayah: Int,
        attemptIndex: Int,
        basename: String,
    ): Map<String, Any?> {
        val modelInfo = stages["modelInfo"] as? Map<*, *>
        val iso = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        return stages.toMutableMap().apply {
            put("surah", surah)
            put("ayah", ayah)
            put("attemptIndex", attemptIndex)
            put("attemptBasename", basename)
            put("expectedRaw", stages["expected"] ?: stages["originalExpectedAyah"] ?: "")
            put("hypRaw", stages["hypothesis"] ?: stages["originalAsrHypothesis"] ?: "")
            put("accuracy", stages["wordAccuracy"] ?: 0.0)
            put("ops", stages["lexicalOps"] ?: emptyList<Any>())
            put("bucket", modelInfo?.get("bucketOrFixedT") ?: stages["bucketT"] ?: 0)
            put("coldLoad", modelInfo?.get("modelWasColdLoad") ?: false)
            put("timestamp", iso.format(Date()))
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun mapToJson(value: Any?): Any {
        return when (value) {
            null -> JSONObject.NULL
            is Map<*, *> -> {
                val obj = JSONObject()
                for ((k, v) in value) obj.put(k.toString(), mapToJson(v))
                obj
            }
            is List<*> -> {
                val arr = JSONArray()
                for (item in value) arr.put(mapToJson(item))
                arr
            }
            is Number, is Boolean, is String -> value
            else -> value.toString()
        }
    }

    fun writeWavMono16k(pcm: FloatArray, file: File) {
        val sampleRate = 16_000
        val dataSize = pcm.size * 2
        val out = ByteArray(44 + dataSize)
        val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        fun ascii(s: String) = s.toByteArray(Charsets.US_ASCII)

        bb.put(ascii("RIFF"))
        bb.putInt(36 + dataSize)
        bb.put(ascii("WAVE"))
        bb.put(ascii("fmt "))
        bb.putInt(16)
        bb.putShort(1) // PCM
        bb.putShort(1) // mono
        bb.putInt(sampleRate)
        bb.putInt(sampleRate * 2)
        bb.putShort(2) // block align
        bb.putShort(16) // bits
        bb.put(ascii("data"))
        bb.putInt(dataSize)
        for (s in pcm) {
            val clamped = s.coerceIn(-1f, 1f)
            bb.putShort((clamped * Short.MAX_VALUE).toInt().toShort())
        }
        FileOutputStream(file).use { it.write(out) }
    }
}
